src/borg/key.py

```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
         return Chunk(data)
 
 
+class NonceManager:
+    def __init__(self, repository, enc_cipher, manifest_nonce):
+        self.repository = repository
+        self.enc_cipher = enc_cipher
+        self.end_of_nonce_reservation = None
+        self.manifest_nonce = manifest_nonce
+        self.key_nonce_file = os.path.join(get_key_nonces_dir(), self.repository.id_str)
+
+    def get_local_free_nonce(self):
+        try:
+            with open(self.key_nonce_file, 'r') as fd:
+                return bytes_to_long(unhexlify(fd.read()))
+        except FileNotFoundError:
+            return None
+
+    def commit_local_nonce_reservation(self, next_unreserved, start_nonce):
+        if self.get_local_free_nonce() != start_nonce:
+            raise Error("nonce space reservation with mismatched previous state")
+        with SaveFile(self.key_nonce_file, binary=False) as fd:
+            fd.write(bin_to_hex(long_to_bytes(next_unreserved)))
+
+    def get_repo_free_nonce(self):
+        try:
+            return self.repository.get_free_nonce()
+        except InvalidRPCMethod as error:
+            # old server version, suppress further calls
+            self.get_repo_free_nonce = lambda: None
+            self.commit_repo_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+
+    def commit_repo_nonce_reservation(self, next_unreserved, start_nonce):
+        self.repository.commit_nonce_reservation(next_unreserved, start_nonce)
+
+    def ensure_reservation(self, nonce_space_needed):
+        if not self.end_of_nonce_reservation:
+            repo_free_nonce = self.get_repo_free_nonce()
+            local_free_nonce = self.get_local_free_nonce()
+            next_nonce = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce) if x is not None)
+            reservation_end = next_nonce + nonce_space_needed + NONCE_SPACE_RESERVATION
+            assert reservation_end < MAX_REPRESENTABLE_NONCE
+            self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            self.enc_cipher.reset(None, next_nonce.to_bytes(16, byteorder='big'))
+        else:
```

Comment:
this order puts initialization before the normal code. But i’m happy to swap it around if you think it does make it more readable.